### ISSUE TITLE
Question: does async_logger::flush() really block?

### ISSUE THREAD

**Issue by arthur-tacca** [gabime/spdlog#930]
According to the Wiki:

> **Important: logger->flush() is always a synchronous operation, even if the logger is in async mode!** In fact, in async mode it might take a long time since spdlog will first wait for the queue to drain before flushing.

But looking at the source code it seems to me that this is out of date. Calling `flush()` on an async logger posts a message to the thread pool queue but I cannot see where it waits for it. Have I missed something? This is what I see:

 1. I call `async_logger::flush()`, which is actually the inherited `logger::flush()`
 2. `logger::flush()`: calls `async_logger::flush_()`
 3. `async_logger::flush_()`: this calls `pool_ptr->post_flush(shared_from_this(), overflow_policy_);` (after acquiring a lock on the thread pool queue).
 4. `thread_pool::post_flush`: calls `post_async_msg_(async_msg(std::move(worker_ptr), async_msg_type::flush), overflow_policy);`
 5. `thread_pool::post_async_msg_`: calls `q_.enqueue_nowait(std::move(new_msg));` (assuming the policy is not to block).

For me this would be a good thing: I want to write every log message to disk but without holding up the non-logging threads, like I could in spdlog 0.16 by combining the async logger with `file_sink::set_force_flush(true)`. (I am using Visual Studio, which I believe is a lot more aggressive about file buffering than libc.) If the above is true then I can just call `async_logger::flush_on(trace)` to get that old behaviour.

Is that right? And if so, is it guaranteed to stay this way in future? To be honest spdlog seems to regularly have breaking changes between releases, and I am worried it could just as easily be changed back to blocking on `flush()`.

**Comment by arthur-tacca** [comment]
Looking further at the source code, I have just noticed that the check for a severity-based flush is actually done in the thread pool (in `async_logger::backend_log_()`), so `async_logger::flush_on(trace)` should definitely not block any other threads. Is that right? If so I'll update the Wiki about both of these things. Again, is this definitely going to stay this way now?

**Comment by gabime** [comment]
Hi,
The wiki is outdated indeed. The flush is done in the thread pool like you noticed.

Of course, I cannot guarantee anything about the future. I don't see a reason why it would change,  but no guarantees.

It would be great If you could fix the wiki.

> a lot more aggressive about file  buffering..

Interesting.. What do you mean?

**Comment by arthur-tacca** [comment]
Thanks for the clarification. I see you have removed the warning that flush() could block from the wiki. I have also added a note explicitly saying that it doesn't block, but feel free to revert if you don't think it's needed.

BTW I think I was wrong saying that Windows is more aggresive about file buffering. It certainly does do file buffering, so without flushing I was seeing partially complete log messages if I looked at logs while the program was running. I was under the impression that Linux is line-buffered for writing text files by default, but now I'm not sure that's the case - I don't have a Linux installation to hand to check.

Digression over - thanks again for the clarification and all your good work on spdlog.
